### PR TITLE
Reimplement

### PR DESCRIPTION
**Pull request by kaidokert**
## Summary by Sourcery

Introduce a new `ed25519_heapless` crate providing generic, no_std-friendly Ed25519 verification over pluggable bigint backends, replace the old bigint-based implementation and binaries with a Montgomery/NAF-based verifier, and wire it into the workspace, tests, and CI.

New Features:
- Expose a generic `verify` API for Ed25519 signatures parameterized over an `UnsignedModularInt` bigint trait with fixed-bigint implementations and associated field/curve constants.
- Add a `verify_baked` binary and a generic `verify` CLI that select concrete fixed-bigint backends via Cargo features for host-side verification.

Bug Fixes:
- Harden signature verification by enforcing strict scalar and point validation, rejecting non-canonical and malleable signatures in line with RFC 8032 and Wycheproof vectors.

Enhancements:
- Reimplement core Ed25519 group arithmetic using Montgomery multiplication, lazy field reductions, and paired-NAF double-scalar multiplication for better performance and backend-independence.
- Modularize the implementation into dedicated modules for scalar recoding (JSF/NAF), lazy field operations, and Montgomery context management.
- Simplify and modernize the keygen and sign binaries by marking them as unimplemented pending constant‑time implementations, focusing the crate on verification.

Build:
- Convert the repository into a Cargo workspace and rename the crate to `ed25519_heapless` with updated metadata, features, and dependencies for fixed-bigint, logging, and hmac-sha512.
- Adjust binary targets and feature gating in Cargo.toml so verification binaries build only when the `fixed-bigint` backend is available.

CI:
- Revise the GitHub Actions workflow to run fmt/clippy for `ed25519_heapless`, exercise no_std builds, run tests (including with fixed-bigint), and perform an integration run of the `verify_baked` binary.

Documentation:
- Replace the README with a brief description of the `ed25519_heapless` fork, its microcontroller focus, and current verification-only scope.

Tests:
- Add comprehensive Ed25519 verification tests including RFC 8032 vectors, Wycheproof-derived positive and negative cases, and regression scenarios to validate strict verification semantics and backend correctness.

Chores:
- Update pre-commit configuration and gitignore entries to reflect the renamed `ed25519_heapless` crate and new workspace layout.

## Summary by CodeRabbit

* **New Features**
  * Adds a full Ed25519 verification implementation and a self-contained "baked" verifier binary.

* **Tests**
  * Adds extensive RFC8032 and Wycheproof test suites plus integration tests for multiple bigint backends.

* **Chores**
  * Crate/workspace renamed and manifest updated; CI schedule and workflow expanded; gitignore and pre-commit entry updated; Makefile removed.

* **Documentation**
  * README simplified to reflect verification-only scope.

* **Breaking Changes**
  * Key-generation and signing CLIs disabled/unimplemented; prior bigint implementation removed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,17 +1,5 @@
...
-- For release : `make all`
-- To run code : `cargo run --bin=[keygen|sign|verify] arg1 arg2 arg3`
-- There are test units in lib.rs for most functions used.
-- Alternatively : you can run the programs found under the folder `release` after having done `make all`
-
-
-# For cleanup
-
-- Run `make clean`
+It only does signature verifications right now ( e.g. bootloaders ) as constant time ops and private key handling requires constant-time ops.
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply README.md]: addressed

File: README.md [hunk 2]
```diff
@@ -1,17 +1,5 @@
-OUERTANI Mohamed Hachem
+### ed25119 heapless fork
 
-# Requirements
+This is a fork of `ed25119` crate that is ported to microcontrollers. It's tested on 8-bit AVR and Cortex-M
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply README.md]: done

File: ed25519/src/jsf.rs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+            s_working >>= 1;
+            h_working >>= 1;
+
+            if s_carry {
+                s_working = s_working + T::one();
+            }
+            if h_carry {
+                h_working = h_working + T::one();
+            }
+        }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply ed25519/src/jsf.rs]: done

File: ed25519/src/strict.rs [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+        let d_owned = d + &T::zero();
+        let y_owned = y + T::zero();
+        let x = recover_x_mont(y, sign, p, d_owned, ctx);
+        x.map(|x_val| {
+            // Compute T = x*y using Montgomery mul, but return point in normal form
+            let x_m = ctx.to_mont(x_val);
+            let y_m = ctx.to_mont(y_owned);
+            let t = ctx.from_mont(ctx.mont_mul(&x_m, &y_m));
+            (x_val, y_owned, T::one(), t)
+        })
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply ed25519/src/strict.rs]: done

File: ed25519/src/strict.rs [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+        match digit.h_digit {
+            1 => result = point_add_niels_mont(&result, &a_niels, p, ctx),
+            -1 => {
+                let neg = (a_niels.1, a_niels.0, lazy_mod_sub(T::zero(), &a_niels.2, p));
+                result = point_add_niels_mont(&result, &neg, p, ctx);
+            }
+            0 => {}
+            _ => unreachable!("JSF digits must be -1, 0, or 1"),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply ed25519/src/strict.rs]: done

File: ed25519/src/bin/keygen.rs [hunk 6]
```diff
@@ -1,46 +1,3 @@
...
-        Ok(mut p) => {
-            p.write_all(&public_key[0..32]).unwrap();
-        }
-
-        Err(e) => {
-            eprintln!("Error creating public key file: {}", e);
-            std::process::exit(1);
-        }
-    }
+    todo!("keygen is not yet implemented")
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply ed25519/src/bin/keygen.rs]: thats okay, it's marked as todo

File: ed25519/src/lib.rs [hunk 7]
```diff
@@ -1,752 +1,152 @@
...
+    + for<'a> core::ops::DivAssign<&'a Self>
+    + for<'a> core::ops::Rem<&'a Self, Output = Self>
+    + for<'a> core::ops::Div<&'a Self, Output = Self>
+    + for<'a> core::ops::Mul<&'a Self, Output = Self>
+    + for<'a> core::ops::Add<&'a Self, Output = Self>
+    + for<'a> core::ops::Sub<&'a Self, Output = Self>
+    + for<'a> core::ops::AddAssign<&'a Self>
+{
+    fn from_bytes_le(bytes: &[u8]) -> Self;
+    fn to_bytes_le<'a>(&self, out: &'a mut [u8]) -> &'a [u8];
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply ed25519/src/lib.rs:61]: done

File: ed25519/src/bin/verify.rs [hunk 8]
```diff
@@ -1,40 +1,33 @@
...
-    let data = std::fs::read(datafile).unwrap();
-    let signature = std::fs::read(sigfile).unwrap();
+    });
 
     let pk: [u8; 32] = pk.try_into().expect("public key is not 32 bytes");
     let signature: [u8; 64] = signature.try_into().expect("signature is not 64 bytes");
 
-    let verification = ed25519::ed25519::verify(pk, &data, signature);
+    let verification =
+        ed25519_heapless::verify::<fixed_bigint::FixedUInt<u32, 16>>(pk, &data, signature);
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply ed25519/src/bin/verify.rs]: done

File: ed25519/src/strict.rs [hunk 9]
```diff
@@ -0,0 +1,507 @@
...
+    // At each step acc < q, so 2*acc < 2q < 2^254, which fits in 256-bit T.
+    let mut acc = T::zero();
+    let one = T::one();
+    // Process from most significant bit (byte 63, bit 7) to least (byte 0, bit 0)
+    // Hash is little-endian: byte 63 is MSB
+    for byte_idx in (0..64).rev() {
+        for bit_idx in (0..8).rev() {
+            // acc = 2*acc + bit, then reduce
+            let (doubled, _overflow) = acc.overflowing_add(&acc);
+            acc = doubled;
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply ed25519/src/strict.rs:41]: done

File: ed25519/src/bin/sign.rs [hunk 10]
```diff
@@ -1,42 +1,3 @@
...
-    let sign_file = std::fs::File::create(sigfile.clone());
-
-    match sign_file {
-        Ok(mut s) => s.write_all(&sig).unwrap(),
-        Err(e) => {
-            eprintln!("Error creating sigfile : {}", e);
-            std::process::exit(1);
-        }
-    }
+    todo!("sign is not yet implemented")
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply ed25519/src/bin/sign.rs]: per design

File: ed25519/src/lib.rs [hunk 11]
```diff
@@ -1,752 +1,152 @@
...
+    100, 142, 78, 234, 102, 101, 118, 139, 215, 15, 95, 135, 103,
+];
+
+// Precomputed modp_sqrt_m1 = 2^((p-1)/4) mod p for Ed25519
+// This is sqrt(-1) mod p used in point decompression
+// Value: b0a00e4a271beec478e42fad0618432fa7d7fb3d99004d2b0bdfc14f8024832b (hex, big-endian)
+pub const MODP_SQRT_M1_BYTES: [u8; 32] = [
+    0xb0, 0xa0, 0x0e, 0x4a, 0x27, 0x1b, 0xee, 0xc4, 0x78, 0xe4, 0x2f, 0xad, 0x06, 0x18, 0x43, 0x2f,
+    0xa7, 0xd7, 0xfb, 0x3d, 0x99, 0x00, 0x4d, 0x2b, 0x0b, 0xdf, 0xc1, 0x4f, 0x80, 0x24, 0x83, 0x2b,
+];
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply ed25519/src/lib.rs:144]: it's correct - all tests pass

File: ed25519/Cargo.toml [hunk 12]
```diff
@@ -1,25 +1,40 @@
...
-# intentionally empty
+[[bin]]
+name = "verify"
+required-features = ["fixed-bigint"]
+
+[[bin]]
+name = "sign"
 
 [[bin]]
 name = "keygen"
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply ed25519/Cargo.toml:40]: No this is intentional

File: ed25519/src/lib.rs [hunk 13]
```diff
@@ -1,752 +1,155 @@
...
+    + core::ops::ShrAssign<usize>
+    + for<'a> core::ops::RemAssign<&'a Self>
+    + for<'a> core::ops::DivAssign<&'a Self>
+    + for<'a> core::ops::Rem<&'a Self, Output = Self>
+    + for<'a> core::ops::Div<&'a Self, Output = Self>
+    + for<'a> core::ops::Mul<&'a Self, Output = Self>
+    + for<'a> core::ops::Add<&'a Self, Output = Self>
+    + for<'a> core::ops::Sub<&'a Self, Output = Self>
+    + for<'a> core::ops::AddAssign<&'a Self>
+{
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply ed25519/src/lib.rs:56]: It is impossible, AFAIK to move bounds like `for<'a> core::ops::Add` - e.g ref bounds - to base trait

File: ed25519/src/strict.rs [hunk 14]
```diff
@@ -0,0 +1,500 @@
...
+            ctx.to_mont(g.2),
+            ctx.to_mont(g.3),
+        )
+    }; // g normal-form (128B) is now dead
+
+    // Phase 5: JSF double scalar multiplication
+    let sbb_minus_haa = jsf_double_scalar_mul_mont(s, &g_mont, h, &neg_aa_mont, &p, &d, &ctx);
+
+    // Phase 6: Final comparison
+    point_equal_mont(&sbb_minus_haa, &rr_mont, &p, &ctx)
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply ed25519/src/strict.rs]: This is not a valid finding for this implementation. Ed25519 as specified in RFC 8032 does not require cofactor multiplication for verification. The standard verification equation is   
  [s]B = R + [H(R,A,M)]A, compared directly without cofactoring. This is the "cofactorless" (or "ZIP-215 compatible") approach used by the original ed25519 ref10 implementation.        
                                                                                                                                                                                           
  Cofactored verification ([8][s]B = [8]R + [8][H]A) is a stricter variant (ed25519ctx/Ed448 style) that rejects more edge cases but is not what standard Ed25519 does. The Wycheproof test
   vectors in the test suite validate this behavior.

**Comment by kaidokert** [comment]
@sourcery-ai review

File: ed25519/Cargo.toml [hunk 15]
```diff
@@ -1,25 +1,40 @@
...
+required-features = ["fixed-bigint"]
 
-[dev-dependencies]
-# intentionally empty
+[[bin]]
+name = "verify"
+required-features = ["fixed-bigint"]
+
+[[bin]]
+name = "sign"
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply ed25519/Cargo.toml:37]: That's intentional

File: ed25519/src/bin/verify.rs [hunk 16]
```diff
@@ -1,44 +1,78 @@
...
+fn main_inner<T>()
+where
+    T: ed25519_heapless::UnsignedModularInt
+        + Copy
+        + modmath::WideMul
+        + modmath::CiosMontMul
+        + num_traits::ops::overflowing::OverflowingAdd
+        + num_traits::WrappingMul
+        + num_traits::WrappingAdd
+        + num_traits::WrappingSub,
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply ed25519/src/bin/verify.rs:10]: Nah, it's fine for now

File: ed25519/src/jsf.rs [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+            let h_bit = (&h_working & &one) == one;
+
+            // NAF recoding rules to minimize non-zero digits
+            let (s_digit, s_carry) = if s_bit {
+                let s_low_2bits = &s_working & &three; // & 3
+                if s_low_2bits == one || s_low_2bits == two {
+                    (1i8, false) // s ≡ 1 or 2 (mod 4)
+                } else {
+                    (-1i8, true) // s ≡ 3 (mod 4)
+                }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply ed25519/src/jsf.rs]: done

**Comment by kaidokert** [comment]
@sourcery-ai review

File: ed25519/tests/ed25519_vectors.rs [hunk 18]
```diff
@@ -0,0 +1,620 @@
+//! Ed25519 test vectors from RFC 8032, Wycheproof, and negative tests
+//!
+//! These tests require the `fixed-bigint` feature.
+//! Run with: cargo test -p ed25519_heapless --features fixed-bigint
+
+#![cfg(feature = "fixed-bigint")]
+
+type T = fixed_bigint::FixedUInt<u32, 16>;
```
Comment (SourceryAI) [inline ed25519/tests/ed25519_vectors.rs:8]: **suggestion (testing):** Tests only exercise a single bigint backend; consider also testing the other supported `fixed-bigint` configurations.

Currently the RFC 8032 and Wycheproof vectors only run with `FixedUInt<u32, 16>`. Since the verifier is meant to support multiple `fixed-bigint` backends (u32x16, u64x8, u64x4, u8x32), please add a way (e.g., cfg-gated duplicate modules or a small macro) to run the same test suite for each enabled backend so backend-specific issues (e.g., limb-width or carry bugs) are caught early.

Suggested implementation:

```rust
#![cfg(feature = "fixed-bigint")]

/// Instantiate Ed25519 test vectors for a specific `fixed-bigint` backend.
///
/// `$modname` is a unique module name for the backend, and `$uint` is the
/// concrete `fixed_bigint::FixedUInt<_, _>` type to use.
macro_rules! ed25519_fixed_bigint_tests {
    ($modname:ident, $uint:ty) => {
        mod $modname {
            use super::*;

            /// Backend-specific big integer type used by these tests.
            type T = $uint;

            // NOTE: The rest of the test suite (helpers and tests that currently
            // rely on `T`) should be moved into this module so they are
            // instantiated once per backend.
        }
    };
}

/// Instantiate the Ed25519 test suite for each supported `fixed-bigint` backend.
/// Only backends whose features are enabled will be compiled and run.
#[cfg(feature = "fixed-bigint-u32x16")]
ed25519_fixed_bigint_tests!(backend_u32x16, fixed_bigint::FixedUInt<u32, 16>);

#[cfg(feature = "fixed-bigint-u64x8")]
ed25519_fixed_bigint_tests!(backend_u64x8, fixed_bigint::FixedUInt<u64, 8>);

#[cfg(feature = "fixed-bigint-u64x4")]
ed25519_fixed_bigint_tests!(backend_u64x4, fixed_bigint::FixedUInt<u64, 4>);

#[cfg(feature = "fixed-bigint-u8x32")]
ed25519_fixed_bigint_tests!(backend_u8x32, fixed_bigint::FixedUInt<u8, 32>);

fn hex_to_bytes(hex: &str) -> Vec<u8> {

```

To fully implement the suggestion and keep the file compiling, you’ll need to:

1. Move the existing helpers and tests that currently use the global `type T = ...` into the macro-generated modules:
   - One simple way:
     - Cut everything **after** the macro (starting from `fn hex_to_bytes` down to the end of the file) and paste it **inside** the `mod $modname { ... }` body, replacing the `// NOTE:` comment.
     - Ensure all `use` statements that those tests rely on are either moved into the module or brought in via `use super::...;`.
2. If you prefer to keep helpers like `hex_to_bytes` at the top level:
   - Leave helper functions (`hex_to_bytes`, `hex_to_array_32`, etc.) at the top level.
   - Inside the macro’s `mod $modname { ... }`, create a nested `mod tests` (or similar) and move the actual `#[test]` functions into it, making sure they import the helpers with `use super::super::*;`.
3. Ensure that the features `fixed-bigint-u32x16`, `fixed-bigint-u64x8`, `fixed-bigint-u64x4`, and `fixed-bigint-u8x32` (or whatever names your crate actually uses for these backends) are defined in `Cargo.toml` and match the `#[cfg(feature = "...")]` gates used here.
4. If your current tests rely on module paths that include `crate::` or `super::`, verify those paths still resolve correctly when the tests are nested inside the macro-generated modules, adjusting `use` statements as needed.

**Review by SourceryAI — COMMENTED** [review]
Hey - I've found 1 issue, and left some high level feedback:

- The `MODP_SQRT_M1_BYTES` constant is documented as a big-endian hex value but is passed into `from_bytes_le`, which expects little-endian; please double-check the endianness of this and other field constants (`P_BYTES`, `D_BYTES`, `Q_BYTES`, `G_*_BYTES`) so they match the deserialization convention or adjust the byte order accordingly.
- The trait bounds for the bigint backend are repeated verbosely across `UnsignedModularInt`, `verify_mont`, `sha512_modq`, `naf_double_scalar_mul_mont`, and the `main_inner` functions; consider introducing a dedicated backend trait or a `type` alias for the combined bounds to reduce duplication and the risk of these drifting out of sync.
- Both `src/bin/verify.rs` and `src/bin/verify_baked.rs` define identical `main_inner<T>` bounds and backend selection cfgs; factoring this common logic into a shared helper (e.g., a small module or function in the library) would make it easier to maintain when backend options or trait requirements change.

<details>
<summary>Prompt for AI Agents</summary>

~~~markdown
Please address the comments from this code review:

## Overall Comments
- The `MODP_SQRT_M1_BYTES` constant is documented as a big-endian hex value but is passed into `from_bytes_le`, which expects little-endian; please double-check the endianness of this and other field constants (`P_BYTES`, `D_BYTES`, `Q_BYTES`, `G_*_BYTES`) so they match the deserialization convention or adjust the byte order accordingly.
- The trait bounds for the bigint backend are repeated verbosely across `UnsignedModularInt`, `verify_mont`, `sha512_modq`, `naf_double_scalar_mul_mont`, and the `main_inner` functions; consider introducing a dedicated backend trait or a `type` alias for the combined bounds to reduce duplication and the risk of these drifting out of sync.
- Both `src/bin/verify.rs` and `src/bin/verify_baked.rs` define identical `main_inner<T>` bounds and backend selection cfgs; factoring this common logic into a shared helper (e.g., a small module or function in the library) would make it easier to maintain when backend options or trait requirements change.

## Individual Comments

### Comment 1
<location path="ed25519/tests/ed25519_vectors.rs" line_range="8" />
<code_context>
+
+#![cfg(feature = "fixed-bigint")]
+
+type T = fixed_bigint::FixedUInt<u32, 16>;
+
+fn hex_to_bytes(hex: &str) -> Vec<u8> {
</code_context>
<issue_to_address>
**suggestion (testing):** Tests only exercise a single bigint backend; consider also testing the other supported `fixed-bigint` configurations.

Currently the RFC 8032 and Wycheproof vectors only run with `FixedUInt<u32, 16>`. Since the verifier is meant to support multiple `fixed-bigint` backends (u32x16, u64x8, u64x4, u8x32), please add a way (e.g., cfg-gated duplicate modules or a small macro) to run the same test suite for each enabled backend so backend-specific issues (e.g., limb-width or carry bugs) are caught early.

Suggested implementation:

```rust
#![cfg(feature = "fixed-bigint")]

/// Instantiate Ed25519 test vectors for a specific `fixed-bigint` backend.
///
/// `$modname` is a unique module name for the backend, and `$uint` is the
/// concrete `fixed_bigint::FixedUInt<_, _>` type to use.
macro_rules! ed25519_fixed_bigint_tests {
    ($modname:ident, $uint:ty) => {
        mod $modname {
            use super::*;

            /// Backend-specific big integer type used by these tests.
            type T = $uint;

            // NOTE: The rest of the test suite (helpers and tests that currently
            // rely on `T`) should be moved into this module so they are
            // instantiated once per backend.
        }
    };
}

/// Instantiate the Ed25519 test suite for each supported `fixed-bigint` backend.
/// Only backends whose features are enabled will be compiled and run.
#[cfg(feature = "fixed-bigint-u32x16")]
ed25519_fixed_bigint_tests!(backend_u32x16, fixed_bigint::FixedUInt<u32, 16>);

#[cfg(feature = "fixed-bigint-u64x8")]
ed25519_fixed_bigint_tests!(backend_u64x8, fixed_bigint::FixedUInt<u64, 8>);

#[cfg(feature = "fixed-bigint-u64x4")]
ed25519_fixed_bigint_tests!(backend_u64x4, fixed_bigint::FixedUInt<u64, 4>);

#[cfg(feature = "fixed-bigint-u8x32")]
ed25519_fixed_bigint_tests!(backend_u8x32, fixed_bigint::FixedUInt<u8, 32>);

fn hex_to_bytes(hex: &str) -> Vec<u8> {

```

To fully implement the suggestion and keep the file compiling, you’ll need to:

1. Move the existing helpers and tests that currently use the global `type T = ...` into the macro-generated modules:
   - One simple way:
     - Cut everything **after** the macro (starting from `fn hex_to_bytes` down to the end of the file) and paste it **inside** the `mod $modname { ... }` body, replacing the `// NOTE:` comment.
     - Ensure all `use` statements that those tests rely on are either moved into the module or brought in via `use super::...;`.
2. If you prefer to keep helpers like `hex_to_bytes` at the top level:
   - Leave helper functions (`hex_to_bytes`, `hex_to_array_32`, etc.) at the top level.
   - Inside the macro’s `mod $modname { ... }`, create a nested `mod tests` (or similar) and move the actual `#[test]` functions into it, making sure they import the helpers with `use super::super::*;`.
3. Ensure that the features `fixed-bigint-u32x16`, `fixed-bigint-u64x8`, `fixed-bigint-u64x4`, and `fixed-bigint-u8x32` (or whatever names your crate actually uses for these backends) are defined in `Cargo.toml` and match the `#[cfg(feature = "...")]` gates used here.
4. If your current tests rely on module paths that include `crate::` or `super::`, verify those paths still resolve correctly when the tests are nested inside the macro-generated modules, adjusting `use` statements as needed.
</issue_to_address>
~~~

</details>

***

Hi @kaidokert! 👋

Thanks for trying out Sourcery by commenting with `@sourcery-ai review`! 🚀

Install the [sourcery-ai bot](https://github.com/apps/sourcery-ai/installations/new) to get automatic code reviews on every pull request ✨

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>
